thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/utils/ThirdeyeMetricsUtil.java

```diff
@@ -76,7 +81,191 @@ private ThirdeyeMetricsUtil() {
   public static final Counter cubeDurationCounter =
       metricsRegistry.newCounter(ThirdeyeMetricsUtil.class, "cubeDurationCounter");
 
+  private static final ConcurrentLinkedDeque<RequestLogEntry> dataSourceRequests = new ConcurrentLinkedDeque<>();
+
   public static MetricsRegistry getMetricsRegistry() {
     return metricsRegistry;
   }
+
+  /**
+   * Append statistics for a successful data source request to the performance log
+   *
+   * @param datasource data source name
+   * @param dataset data set name
+   * @param metric metric name
+   * @param start request start time in ns
+   * @param end request end time in ns
+   */
+  public static void logRequestSuccess(String datasource, String dataset, String metric, long start, long end) {
+    dataSourceRequests.add(new RequestLogEntry(datasource, dataset, metric, getPrincipal(), true, start, end, null));
```

Comment:
Set a size limitation or truncate the log on-the-fly to prevent the queue consumes all memory if the machine is not rebooted for a long time or there are tons of requests in a very short time. Remember that each request also makes data cache bigger.

Reply:
The queue is drained every hour by the logger. Counting elements isn't O(1) because it's a concurrent structure. In the best case we can have an approximate gauge.